### PR TITLE
More flexible way to create and test forests

### PR DESCRIPTION
**Pull request by adrianludwin**
I removed some of the tests about missing intermediate namespaces since they're not actually possible anymore, and my next PR will remove the ability to express them.

### CODE REVIEW

File: incubator/hnc/pkg/validators/hierarchy_test.go [hunk 1]
```diff
@@ -10,16 +10,11 @@ import (
...
 )
 
 func TestStructure(t *testing.T) {
-	f := forest.NewForest()
-	// Create the following forest: foo -> bar
-	foo := createNS(f, "foo", nil)
-	bar := createNS(f, "bar", nil)
-	createNS(f, "baz", nil)
-	bar.SetParent(foo)
+	f := foresttest.Create("-a-") // a -> b; c
```
Comment (yiqigao217) [inline incubator/hnc/pkg/validators/hierarchy_test.go:17]: nit: s/a->b/a<-b to match what in the `foresttest.Create` examples (arrow pointing to the parent).

Comment (adrianludwin) [reply incubator/hnc/pkg/validators/hierarchy_test.go]: Good catch, fixed

**Comment by yiqigao217** [comment]
> I removed some of the tests about missing intermediate namespaces since they're not actually possible anymore, and my next PR will remove the ability to express them.

Question: why missing intermediate namespaces was possible before and is not possible now?

**Comment by adrianludwin** [comment]
> Question: why missing intermediate namespaces was possible before and is not possible now?

Before, we could have namespace `A` with requiredChild `B`, and namespace `C` could have parent `B`, but there could be some problem that prevented `B` from being created. That is, `B` is an intermediate namespace between `C` and `A`, and it would appear in the in-memory forest as both a child of `A` and a parent of `C`, just with `ns.Exists()` returning false.

But now, when we create an anchor in `A`, we don't actually create an entry for `B` in the in-memory forest until the Anchor reconciler successfully creates the namespace. So `B` will never appear to have parent `A` until it actually gets created.

By contrast, if `C` reference parent `B` and it doesn't exist, we _will_ create it in-memory. But it won't appear to be an _intermediate_ namespace anymore - it will just look like a root.

Does this make sense?

**Comment by yiqigao217** [comment]
I see makes sense, thanks for the explanation!

/lgtm
/approve
/hold cancel

**Pull request merged**